File: starter_kits/Ruby/hlt/ship.rb
require 'entity'
require 'commands'
require 'constants'
require 'direction'
require 'position'


# Ship class to house ship entities
class Ship < Entity
  attr_reader :halite_amount

  def initialize(owner, id, position, halite_amount)
    super(owner, id, position)
    @halite_amount = halite_amount
  end

  # Is this ship at max halite capacity?
  def is_full?
    @halite_amount >= Constants::MAX_HALITE
  end

  # Return a move to transform this ship into a dropoff.
  def make_dropoff
    "#{Commands::CONSTRUCT} #{@id}"
  end

  # Return a move to move this ship in a direction without
  # checking for collisions.
  def move(direction)
    raw_direction = direction
    if !direction.is_a?(String) || !"nsewo".include?(direction)
      raw_direction = Direction.convert(direction)
    end
    "#{Commands::MOVE} #{@id} #{raw_direction}"
  end

  # Don't move this ship.
  def stay_still
    "#{Commands::MOVE} #{@id} #{Commands::STAY_STILL}"
  end

  # Creates an instance of a ship for a given player given the engine's input.
  # :param player_id: The id of the player who owns this ship
  # :param player_id: The id of the player who owns this ship
  # :return: The ship id and ship object
  def self.generate(game, player_id)
    id, x, y, halite = game.read_ints_from_input
    Ship.new(player_id, id, Position.new(x, y), halite)
  end

  def to_s
    "#{self.class}(id=#{@id}, #{@position}, cargo=#{@halite_amount} halite)"
  end
end

File: starter_kits/TypeScript/hlt/Ship.ts
import { Commands } from './commands';
import { Direction } from "./Direction";
import { Position } from "./Position";
import { Constants } from './constants';
import { Entity } from './entity';

/** Represents a ship. */
export class Ship extends Entity {
    constructor(owner: number, id: number, position: Position, public haliteAmount: number) {
        super(owner, id, position);
    }

    /** Is this ship at max halite capacity? */
    get isFull() {
        return this.haliteAmount >= Constants.MAX_ENERGY;
    }

    /** Return a move to turn this ship into a dropoff. */
    makeDropoff() {
        return `${Commands.CONSTRUCT} ${this.id}`;
    }

    /**
     * Return a command to move this ship in a direction without
     * checking for collisions.
     */
    move(direction: Direction) {
        return `${Commands.MOVE} ${this.id} ${direction.toWireFormat()}`;
    }

    /**
     * Return a command to not move this ship.
     *
     * Not strictly needed, since ships do nothing by default.
     */
    stayStill() {
        return `${Commands.MOVE} ${this.id} ${Commands.STAY_STILL}`;
    }

    updateShip(x: number, y: number, halite: number) {
        this.position.x = x;
        this.position.y = y;
        this.haliteAmount = halite;
    }

    toString() {
        return `${this.constructor.name}(id=${this.id}, ${this.position}, cargo=${this.haliteAmount} halite)`;
    }
}